hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -4815,3 +4887,206 @@ func isAPIServerRoute(hcluster *hyperv1.HostedCluster) bool {
+						Ports:     []corev1.ContainerPort{{Name: "metrics", Protocol: "TCP", ContainerPort: 8443}},
+						Resources: cpoResources,
+					},
+				},
+			},
+		},
+	}
+
+	if openShiftTrustedCABundleConfigMapExists {
+		hyperutil.DeploymentAddOpenShiftTrustedCABundleConfigMap(deployment)
+	}
+
+	mainContainer = hyperutil.FindContainer("control-plane-pki-operator", deployment.Spec.Template.Spec.Containers)
+	proxy.SetEnvVars(&mainContainer.Env)
+
+	// set security context
+	if setDefaultSecurityContext {
+		deployment.Spec.Template.Spec.SecurityContext = &corev1.PodSecurityContext{
+			RunAsUser: k8sutilspointer.Int64(config.DefaultSecurityContextUser),
+		}
+	}
+
+	deploymentConfig := config.DeploymentConfig{
+		Scheduling: config.Scheduling{
+			PriorityClass: config.DefaultPriorityClass,
+		},
+		SetDefaultSecurityContext: setDefaultSecurityContext,
+		AdditionalLabels: map[string]string{
+			config.NeedManagementKASAccessLabel: "true",
+		},
+	}
+	if hcp.Annotations[hyperv1.ControlPlanePriorityClass] != "" {
+		deploymentConfig.Scheduling.PriorityClass = hcp.Annotations[hyperv1.ControlPlanePriorityClass]
+	}
+	deploymentConfig.SetDefaults(hcp, nil, k8sutilspointer.Int(1))
+	deploymentConfig.SetRestartAnnotation(hc.ObjectMeta)
+	deploymentConfig.ApplyTo(deployment)
+
+	return nil
+}
+
+func reconcileControlPlanePKIOperatorRole(role *rbacv1.Role) error {
+	role.Rules = []rbacv1.PolicyRule{
+		{ // to create owner refs
+			APIGroups: []string{"hypershift.openshift.io"},
+			Resources: []string{"hostedcontrolplanes"},
+			Verbs:     []string{"get"},
+		},
+		{ // to report status
+			APIGroups: []string{"hypershift.openshift.io"},
+			Resources: []string{"hostedcontrolplanes/status"},
+			Verbs:     []string{"patch"},
+		},
+		{ // to do the work of the controller
+			APIGroups: []string{""},
+			Resources: []string{"configmaps", "secrets", "events"},
+			Verbs:     []string{"get", "list", "watch", "create", "delete", "update", "patch"},
+		},
+		{ // for owner reference resolution
+			APIGroups: []string{""},
+			Resources: []string{"pods"},
+			Verbs:     []string{"get"},
+		},
+		{ // for owner reference resolution
+			APIGroups: []string{"apps"},
+			Resources: []string{"replicasets"},
+			Verbs:     []string{"get"},
+		},
+		{ // for leader election
+			APIGroups: []string{"coordination.k8s.io"},
+			Resources: []string{"leases"},
+			Verbs:     []string{"get", "list", "watch", "create", "delete", "update", "patch"},
+		},
+	}
+	return nil
+}
+
+func reconcileControlPlanePKIOperatorRoleBinding(binding *rbacv1.RoleBinding, role *rbacv1.Role, sa *corev1.ServiceAccount) error {
+	binding.RoleRef = rbacv1.RoleRef{
+		APIGroup: "rbac.authorization.k8s.io",
+		Kind:     "Role",
+		Name:     role.Name,
+	}
+
+	binding.Subjects = []rbacv1.Subject{
+		{
+			Kind:      "ServiceAccount",
+			Name:      sa.Name,
+			Namespace: sa.Namespace,
+		},
+	}
+
+	return nil
+}
+
+func reconcileControlPlanePKIOperatorClusterRole(role *rbacv1.ClusterRole) error {
```

Comment:
I'm not understanding the need for this.  Why does the controller need to know the infrastructure of the MC?

Reply:
https://github.com/openshift/hypershift/pull/3193#discussion_r1399742409